crates/biome_js_analyze/src/semantic_analyzers/correctness/use_hook_at_top_level.rs

```diff
@@ -100,14 +121,159 @@ fn enclosing_function_if_call_is_at_top_level(call: &JsCallExpression) -> Option
         .and_then(|body| body.parent::<AnyJsFunction>())
 }
 
+/// Model for tracking which function calls are preceeded by an early return.
+#[derive(Clone, Default)]
+struct EarlyReturnsModel(HashMap<JsCallExpression, TextRange>);
```

Comment:
We usually try to use `FxHashMap`

Reply:
I believe we don't need to track `JsCallExpression`. By looking at the code, it seems we only need to track the range of the call expression. This means that we can use a `FxHashSet<TextRange>`.

And then below, the code will be:

```rs
if let Some(early_return_range) = early_returns.get(call.range()) {
    return Some(Suggestion::None {
        hook_name_range,
        path,
        early_return: Some(early_return_range.clone()),
    });
}
```

Reply:
I'll change to use a `FxHashMap` :+1:

We can't use a set though, because the range of the return isn't the same as the range of the call. We need to determine for every call site, whether there was an early return *prior to* that call site.

Reply:
The description of the model is a bit cryptic, though, because it seems that a function call is associated to **one** early return, but now it seems that we can have two function calls with the same text range. Is that correct?

Would it be possible to expand the model's description to understand better how this data structure is handled? I believe there's room for optimisation

Reply:
Every function call is associated with zero or one early return, that's correct. But yes, it's also possible that multiple function calls get associated with the same early return, so in that case the text range for such early returns would be cloned multiple times. Do you think that would be significant? I suppose it's feasible to wrap them in an `Rc` or `Arc`.

Reply:
I see a `TextRange` is only 8 bytes and derives `Copy`, so I don't think we'd win anything by wrapping it.

Reply:
Also pushed a bit of extra documentation.